Commit with new futures (fixes #5)

mkgerasimenko opened the pull request:
- added Lombok library;
 - added softAssertions;
 - created a custom annotation;

mkgerasimenko committed:
Added and changed the following:
 - added gradle quality plugin;
 - fixed all violations;
 - added TestNG listener;
 - added test-data-supplier library;
 - created new tests;
 - changed gradle version up to the newer.

mkgerasimenko committed:
Added and changed the following:
 - added sample javadoc for both main and test packages;
 - added one more TestNG listener;
 - common refactoring.

mkgerasimenko committed:
Added and changed the following:
 - added Lombok library;
 - refactor entities with new annotations;
 - added softAssertions;
 - created a custom annotation;
 - replaced common java logger with @Slf4j;
 - common refactoring.

sskorol commented on src/test/java/com/waverleysoftware/BaseTest.java:
```diff
@@ -1,16 +1,39 @@
+@Slf4j
 public class BaseTest {
+
+    protected static final SoftAssertions SOFTASSERT = new SoftAssertions();
```
according to Java conventions, each word in a constant should be splitted with `_` character

sskorol commented on src/test/java/com/waverleysoftware/BaseTest.java:
```diff
@@ -1,16 +1,39 @@
+    @BeforeMethod
+    public void setUp(final ITestContext context, final Method method) {
+        log.info(method.getName());
+        log.info(String.valueOf(method.getDeclaredAnnotation(Test.class)));
```
use special `{}` placeholder for injecting variables into logger instance

sskorol commented on src/test/java/com/waverleysoftware/BaseTest.java:
```diff
@@ -1,16 +1,39 @@
     public void tearDown() {
-        System.out.println("Ending process on parent level");
+        log.info("Ending process on parent level");
+        SOFTASSERT.assertAll();
```
it's not recommended to throw errors within configuration methods... try to move this staff into listener, catch an assertion error and manually change test result to failure...

sskorol commented on src/test/java/com/waverleysoftware/Name.java:
```diff
@@ -0,0 +1,18 @@
+import java.lang.annotation.Target;
+
+/**
+ * Name annotation.
```
give move obvious explanation about the purpose of this annotation

sskorol commented on src/test/java/com/waverleysoftware/SensorTests.java:
```diff
@@ -1,20 +1,29 @@
+
+    @Test(dataProvider = "getSensorData", dataProviderClass = DataSuppliers.class)
+    public void sensorShouldBeAdded(final Sensor... sensors) {
+        SOFTASSERT.assertThat(sensors[0].getType().toString().toLowerCase()).as("Sensor type").isEqualTo("embedded");
```
it's much easier to compare type with enum value directly to avoid redundant conversions... besides that, you've mixed expected with actual... `sensors` arrays is passed via data supplier, so it's smth you expect to get as a final result...

sskorol reviewed (changes requested):
@mkgerasimenko check comments please

mkgerasimenko committed:
Changed the following:
 - deleted unused custom annotation;
 - added special {} placeholder for injecting variables into logger instance;
 - moved softAssert to Listener and created get method;
 - common refactoring.

sskorol commented on src/main/java/com/waverleysoftware/listemers/TestListener.java:
```diff
@@ -1,31 +1,41 @@
+@SuppressWarnings({"JavadocType", "unchecked"})
 @Slf4j
 public class TestListener implements IInvokedMethodListener {
+    private static ThreadLocal thread = new ThreadLocal();
```
make it `final`, and specify an exact type, you'll be working with as generic argument

sskorol commented on src/main/java/com/waverleysoftware/listemers/TestListener.java:
```diff
@@ -1,31 +1,41 @@
+import org.testng.internal.TestResult;
 
-@SuppressWarnings("JavadocType")
+@SuppressWarnings({"JavadocType", "unchecked"})
```
avoid using `unchecked` on class level... annotate exact method, where it's required... however, I believe that type casting warning will go away, when you specify an exact generic type

sskorol commented on src/main/java/com/waverleysoftware/listemers/TestListener.java:
```diff
@@ -1,31 +1,41 @@
     @Override
     public void beforeInvocation(final IInvokedMethod method, final ITestResult testResult) {
+        final SoftAssertions softAssert = new SoftAssertions();
+        thread.set(softAssert);
```
try to give it more obvious name, as it's not a thread... it's a thread local container for some object you're passing to it...

sskorol commented on src/main/java/com/waverleysoftware/listemers/TestListener.java:
```diff
@@ -1,31 +1,41 @@
+            try {
+                getSoftAssert().assertAll();
+            } catch (AssertionError e) {
+                testResult.setStatus(TestResult.FAILURE);
```
besides marking test as failed, you have to remove it from passing collection...

sskorol commented on src/main/java/com/waverleysoftware/listemers/TestListener.java:
```diff
@@ -1,31 +1,41 @@
     }
+
+    public static SoftAssertions getSoftAssert() {
+        return (SoftAssertions) thread.get();
```
it's not required if you set an exact generic type on thread local level

sskorol reviewed (changes requested):
@mkgerasimenko check comments please

mkgerasimenko committed:
Changed the following:
 - set an exact generic type on thread local level;
 - changed the variable name;
 - common refactoring

sskorol commented on src/main/java/com/waverleysoftware/listemers/TestListener.java:
```diff
@@ -7,24 +7,26 @@
             try {
                 getSoftAssert().assertAll();
             } catch (AssertionError e) {
+                testResult.getTestContext().getFailedTests().removeResult(testResult.getMethod());
```
you have to remove current test not from failed, but from passed collection

sskorol commented on src/main/java/com/waverleysoftware/listemers/TestListener.java:
```diff
@@ -7,24 +7,26 @@
     public void beforeInvocation(final IInvokedMethod method, final ITestResult testResult) {
         final SoftAssertions softAssert = new SoftAssertions();
-        thread.set(softAssert);
+        THREAD_LOCAL_CONTAINER_FOR_SOFT_ASSERTIONS.set(softAssert);
```
no need to create a separate `softAssert` variable... you can inline it with setter

sskorol commented on src/test/java/com/waverleysoftware/AccessCodeTests.java:
```diff
@@ -15,16 +15,19 @@
     public void accessCodeShouldBeAdded(final Panel... panels) {
         final String conditionForType = "home";
-        getSoftAssert().assertThat(panels[0].getType()).as("Panel Name").isEqualTo(conditionForType);
+        getSoftAssert().assertThat(panels[0].getType())
```
as you're using only first element in all test, it'd make sense to pass only single `Panel` instance...

sskorol commented on src/test/java/com/waverleysoftware/AccessCodeTests.java:
```diff
@@ -15,16 +15,19 @@
         final String conditionForType = "home";
-        getSoftAssert().assertThat(panels[0].getType()).as("Panel Name").isEqualTo(conditionForType);
+        getSoftAssert().assertThat(panels[0].getType())
+                .as("Panel Name").isEqualTo(conditionForType);
```
you've mixed up expected and actual

sskorol reviewed (changes requested):
@mkgerasimenko check comments please

mkgerasimenko committed:
Changed the following:
 - fixed the comments;
 - common refactoring.
